### ISSUE TITLE
[Troubleshoot]: The KL divergence of PPO is unstable

### ISSUE THREAD

**Issue by april-zmx** [RLinf/RLinf#470]
### Problem description

I noticed that when training the PPO algorithm, the values of the KL divergence are generally on the high side and exhibit severe oscillations—it often exceeds or frequently approaches the threshold of 0.01. I used the pi0.5 model for training, and found that after each rollout, the KL divergence is not theoretically 0.0 at the first parameter update; instead, it starts off with a relatively large value, such as reaching 0.01 or even higher. Only by recalculating the KL divergence in the actor worker can an initial value of 0.0 be obtained.

So far, I have conducted training experiments on three types of GPUs: A100, V100, and 4090D, and similar issues have occurred in all cases. However, I observed that switching to a smaller-scale model seems to alleviate this phenomenon.

Have you encountered such problems during your experiments? Because when I referred to the papers you published, I found that the reported KL divergence curves all showed stable and normal trends.

**Comment by chenkang455** [comment]
Hi @april-zmx, thanks for your great question!

Actually, during our early experiments, we also observed that the KL divergence is not theoretically 0.0 at the first parameter update. This is theoretically expected for algorithms like PPO/GRPO, but in practice, a small discrepancy exists. The core reason is that we apply .sum() to the chunk-level actions output by the pi model (e.g., summing over a 7D action with a chunk length of 5), which amplifies numerical errors. You could try replacing .sum() with .mean(), which brings the initial KL div closer to 0; however, this tends to degrade model convergence.

Nonetheless, based on our observations, this issue does not affect training stability. It is normal for the KL divergence to stay around 0.01. Could you share your training configuration and curves?

We recommend using pi0-libero for your experiments, which we have re-validated multiple times. Since pi05-libero already shows high initial performance across spatial, object, and goal tasks, it is more prone to oscillations and requires a smaller learning rate and noise level.

**Comment by april-zmx** [comment]
@chenkang455 I adjusted the overall settings to slightly reduce the batch size. Before recalculating the log probabilities, my training was consistently unstable, with large fluctuations in `success_once` or only negligible improvements.

![Image](https://github.com/user-attachments/assets/4cb3e82e-2279-45c3-8bc3-eed2f31e267c)

![Image](https://github.com/user-attachments/assets/73fdc1ea-ca9e-4f99-9117-5ac7bfe5fb55)

![Image](https://github.com/user-attachments/assets/4b894aea-fe49-44fc-a0cc-636e9651a4d3)

git commit：

![Image](https://github.com/user-attachments/assets/1835e481-aa7d-4539-b50b-a1200dc0b6a1)

**Comment by chenkang455** [comment]
Hi @april-zmx, thanks for sharing your configs.

Regarding "recalculating log probabilities," do you mean recomputing the old policy's log-probs during the actor training stage? Does this improve training stability? (This is an interesting finding!)

As for the config, I noticed you reduced num-envs from 512 (64 * 8) to 32 (32 * 1). A small number of environments per rollout can lead to fluctuations in the success-once metric. I suggest increasing the rollout-epoch to 8 while keeping the global batch size unchanged. If you encounter OOM, you can decrease the micro-batch size instead. Recommended training config:

```
num_group_envs: 32
rollout_epoch: 8/16 (16 so that the total envs are equal with our previous setting 512)
global_batch_size: 2048
micro_batch_size: 32
```

Overall, I think the main reason is that there are few num-envs per each rollout, but your solution seems very fancy! Looking forward to your reply!

**Comment by chenkang455** [comment]
@april-zmx Furthermore, I wonder if you use the same training config for these two experiments? If so, this might be a potential bug for our experiments. Thanks for pointing it out.

**Comment by april-zmx** [comment]
@chenkang455 This is the code where I recalculate the log probabilities (logprobs), which is re-computed in each respective worker before the start of every training run.

![Image](https://github.com/user-attachments/assets/bf14e3bb-f8b5-4b96-9286-5247783116a7)

I have adopted the suggested settings, removed the trick of recalculating log probabilities, and the training is currently in progress, which may take some time.

Will there be baseline results for common RL training tasks later (excluding embodied manipulation tasks like Gym)? For example, results similar to stable_baselines, with smaller models and shorter training cycles.

**Comment by chenkang455** [comment]
> @chenkang455 This is the code where I recalculate the log probabilities (logprobs), which is re-computed in each respective worker before the start of every training run.
> 
> ![Image](https://github.com/user-attachments/assets/bf14e3bb-f8b5-4b96-9286-5247783116a7)
> 
> I have adopted the suggested settings, removed the trick of recalculating log probabilities, and the training is currently in progress, which may take some time.
> 
> Will there be baseline results for common RL training tasks later (excluding embodied manipulation tasks like Gym)? For example, results similar to stable_baselines, with smaller models and shorter training cycles.

Thanks for providing that! And i will try to reproduce your config on my machine in the meantime. For small models, we have supported training a mlp on the maniskill environment. You can find it in the doc. For gym tasks, we will consider it and thanks for your advice. Besides, i want to confirm that you add this re compute logprob trick and the training becomes stable? This will be a usee insight for us to debug on it. Thx a lot on your issue!

**Comment by chenkang455** [comment]
Hi, @april-zmx I have conducted another experiment to simulate your case, i.e., training the VLA with less rollout epochs and envs. I made the following modifications:

```
rollout_epoch: 8 -> 2
total_num_envs: 64 -> 32
micro_batch_size: 128 -> 64
global_batch_size: 2048 -> 256
lr: 5e-6 -> 1e-6
critic_warmup_steps: 0 # optional 0,5,10, if you observe performance drop at first
```

I conduct the experiments with 4 H100 GPUs and the experiment comparison is shown below:

<img width="1442" height="676" alt="Image" src="https://github.com/user-attachments/assets/2d345212-b552-4021-b59c-ff3e1fcbc622" />

This is not the experiment for LIBERO, but I think their effect is common.

**Comment by april-zmx** [comment]
@chenkang455 
1. Yes, I recalculated the likelihood to stabilize the training process, and the overall change in KL divergence has also become much smoother.
2. Thank you for providing the configuration. I am currently conducting training based on this configuration and will share the results later.

**Comment by chenkang455** [comment]
@april-zmx Thanks for that 😊😊😊. Looking forward to your result.

**Comment by chenkang455** [comment]
Hi, @april-zmx I have experimented on the Libero-Spatial, fyi:

<img width="2102" height="658" alt="Image" src="https://github.com/user-attachments/assets/80d098f6-5286-4ff4-89be-547f43a4dc86" />

**Comment by april-zmx** [comment]
> Hi, [@april-zmx](https://github.com/april-zmx) I have conducted another experiment to simulate your case, i.e., training the VLA with less rollout epochs and envs. I made the following modifications:
> 
> ```
> rollout_epoch: 8 -> 2
> total_num_envs: 64 -> 32
> micro_batch_size: 128 -> 64
> global_batch_size: 2048 -> 256
> lr: 5e-6 -> 1e-6
> critic_warmup_steps: 0 # optional 0,5,10, if you observe performance drop at first
> ```
> 
> I conduct the experiments with 4 H100 GPUs and the experiment comparison is shown below:
> 
> <img alt="Image" width="1442" height="676" src="https://private-user-images.githubusercontent.com/72788314/529109697-2d345212-b552-4021-b59c-ff3e1fcbc622.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjY2NDQzNTEsIm5iZiI6MTc2NjY0NDA1MSwicGF0aCI6Ii83Mjc4ODMxNC81MjkxMDk2OTctMmQzNDUyMTItYjU1Mi00MDIxLWI1OWMtZmYzZTFmY2JjNjIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMjI1VDA2MjczMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU2NzU3NTBkNGY5NmZlZDViZmQ5NmQ4YzMzOThhNmQ1M2RmNmNkZTM3MjQ1YTgzYTY1YTFjMmYxZTk0ZDYxOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.RtbG1It4BdnceClJ5Y6UuHfEwiY8bFv_lUgEF4s8mQM">
> This is not the experiment for LIBERO, but I think their effect is common.

@chenkang455 
The performance of this set of parameters is indeed excellent, and the training on my side is also very stable.

![Image](https://github.com/user-attachments/assets/443d420b-fdc2-4b33-b8ff-0a4618e8a345)
​
![Image](https://github.com/user-attachments/assets/de125b51-ac1e-499f-9cf4-a905a0437d1c)

**Comment by chenkang455** [comment]
@april-zmx Glad to hear that! And your insights on kl are interesting, we will dig the reason behind it. Thanks for your bug feedback!
